File: src/main/java/ac/dnd/server/common/validator/StartBeforeEndValidator.java
package ac.dnd.server.common.validator;

import java.time.LocalDateTime;

import org.springframework.beans.BeanWrapperImpl;

import ac.dnd.server.common.annotation.StartBeforeEnd;
import jakarta.validation.ConstraintValidator;
import jakarta.validation.ConstraintValidatorContext;

public class StartBeforeEndValidator implements ConstraintValidator<StartBeforeEnd, Object> {

	private String startDateTimeFiledName;
	private String endDateTimeFiledName;

	@Override
	public void initialize(final StartBeforeEnd constraintAnnotation) {
		this.startDateTimeFiledName = constraintAnnotation.startDateTime();
		this.endDateTimeFiledName = constraintAnnotation.endDateTime();
	}

	@Override
	public boolean isValid(
		final Object value,
		final ConstraintValidatorContext context
	) {
		if (value == null) {
			return true;
		}
		LocalDateTime startDateTime = getLocalDateTimeProperty(
			value,
			startDateTimeFiledName
		);
		LocalDateTime endDateTime = getLocalDateTimeProperty(
			value,
			endDateTimeFiledName
		);

		if (startDateTime == null || endDateTime == null) {
			return true;
		}

		boolean isValid = startDateTime.isBefore(endDateTime);

		if (!isValid) {
			context.disableDefaultConstraintViolation();
			context.buildConstraintViolationWithTemplate(context.getDefaultConstraintMessageTemplate())
				.addPropertyNode(endDateTimeFiledName)
				.addConstraintViolation();
		}

		return isValid;
	}

	private LocalDateTime getLocalDateTimeProperty(
		final Object value,
		final String fieldName
	) {
		Object propertyValue = new BeanWrapperImpl(value).getPropertyValue(fieldName);

		if (propertyValue == null) {
			return null;
		}

		if (!(propertyValue instanceof LocalDateTime)) {
			throw new IllegalArgumentException(String.format(
				"필드 '%s'의 타입이 LocalDateTime이어야 합니다.",
				fieldName
			));
		}
		return (LocalDateTime)propertyValue;
	}

}

File: src/main/kotlin/ac/dnd/server/global/validator/StartBeforeEndValidator.kt
package ac.dnd.server.global.validator

import ac.dnd.server.global.annotation.StartBeforeEnd
import jakarta.validation.ConstraintValidator
import jakarta.validation.ConstraintValidatorContext
import org.springframework.beans.BeanWrapperImpl
import java.time.LocalDateTime

class StartBeforeEndValidator : ConstraintValidator<StartBeforeEnd, Any> {

    private lateinit var startDateTimeFieldName: String
    private lateinit var endDateTimeFieldName: String

    override fun initialize(constraintAnnotation: StartBeforeEnd) {
        this.startDateTimeFieldName = constraintAnnotation.startDateTime
        this.endDateTimeFieldName = constraintAnnotation.endDateTime
    }

    override fun isValid(value: Any?, context: ConstraintValidatorContext): Boolean {
        if (value == null) {
            return true
        }

        val startDateTime = getLocalDateTimeProperty(value, startDateTimeFieldName)
        val endDateTime = getLocalDateTimeProperty(value, endDateTimeFieldName)

        if (startDateTime == null || endDateTime == null) {
            return true
        }

        val isValid = startDateTime.isBefore(endDateTime)

        if (!isValid) {
            context.disableDefaultConstraintViolation()
            context.buildConstraintViolationWithTemplate(context.defaultConstraintMessageTemplate)
                .addPropertyNode(endDateTimeFieldName)
                .addConstraintViolation()
        }

        return isValid
    }

    private fun getLocalDateTimeProperty(value: Any, fieldName: String): LocalDateTime? {
        val propertyValue = BeanWrapperImpl(value).getPropertyValue(fieldName)

        if (propertyValue == null) {
            return null
        }

        if (propertyValue !is LocalDateTime) {
            throw IllegalArgumentException("필드 '$fieldName'의 타입이 LocalDateTime이어야 합니다.")
        }

        return propertyValue
    }
}